lib/rpc.ts

```
@@ -104,58 +106,87 @@ export class Rpc extends EventEmitter implements IForwarderDest {
-   * to register implementations and add "message" listeners without the risk of missing messages,
-   * even if receiveMessage() has already started being called.
-   *
-   * If sendMessage throws a synchronous exception while processing queued messages, then start()
-   * throws as well, and Rpc remains stopped.
+   * If you've set up calls to receiveMessage(), but need time to call registerImpl() before
+   * processing new messages, you may use queueIncoming(), make the registerImpl() calls,
+   * and then call processIncoming() to handle queued messages and resume normal processing.
+   */
+  public queueIncoming() {
+    if (!this._inactiveRecvQueue) {
+      this._inactiveRecvQueue = [];
+    }
+  }
+
+  /**
+   * Process received messages queued since queueIncoming, and resume normal processing of
+   * received messages.
    */
-  public start(sendMessage: SendMessageCB) {
-    // Message sent by `_dispatch(...)` are appended to the send queue
-    processQueue(this._inactiveRecvQueue, this._dispatch.bind(this));
-    // If send triggers a receive, that would be queued and never processed unless we set
-    // sendMessage now. We reset it on exception.
+  public processIncoming() {
+    if (this._inactiveRecvQueue) {
+      processQueue(this._inactiveRecvQueue, this._dispatch.bind(this));
+      this._inactiveRecvQueue = null;
+    }
+  }
+
+  /**
+   * Set the callback to send messages. If set to null, sent messages will be queued. If you
+   * disconnect and want for sent messages to throw, set a callback that throws.
+   */
+  public setSendMessage(sendMessage: SendMessageCB|null) {
     this._sendMessageCB = sendMessage;
-    try {
-      processQueue(this._inactiveSendQueue, this._sendMessageOrReject.bind(this, sendMessage));
-    } catch (e) {
-      this._sendMessageCB = null;
-      throw e;
+    if (this._sendMessageCB) {
+      try {
+        this._processOutgoing();
```

How about instead doing `this.emit("error", e)` in _sendReject, just before it re-throws? This way the error-emitting would happen in all cases from one place. Then here we won't need try/catch wrap at all, and only in MsgType.Ready handler, we'll need a try/catch that swallows the error silently.